SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc

```diff
@@ -119,39 +123,33 @@ void HGCalSiNoiseMap::setGeometry(const CaloSubdetectorGeometry *hgcGeom, GainRa
 }
 
 //
-const HGCalSiNoiseMap::SiCellOpCharacteristicsCore HGCalSiNoiseMap::getSiCellOpCharacteristicsCore(
-    const HGCSiliconDetId &cellId, GainRange_t gain, int aimMIPtoADC) {
+template <typename T>
+const typename HGCalSiNoiseMap<T>::SiCellOpCharacteristicsCore HGCalSiNoiseMap<T>::getSiCellOpCharacteristicsCore(
+    const T &cellId, GainRange_t gain, int aimMIPtoADC) {
   //re-compute
   if (!activateCachedOp_)
     return getSiCellOpCharacteristics(cellId, gain, aimMIPtoADC).core;
 
-  //re-use from cache
-  HGCSiliconDetId posCellId(cellId.subdet(),
-                            1,
-                            cellId.type(),
-                            cellId.layer(),
-                            cellId.waferU(),
-                            cellId.waferV(),
-                            cellId.cellU(),
-                            cellId.cellV());
-  uint32_t key(posCellId.rawId());
+  uint32_t key = cellId.rawId();
```

Comment:
would be good if @franzoni @pfs confirm that is correct and why was done like that before

Reply:
Ciao @mariadalfonso, it does not seem correct to me, but I'm sure @franzoni and @pfs can comment more.
IIUC, you fill the cache using only 1 side of the detector, and you play the trick while querying the cache by "forcing" the side in the `detId` while you build it. Unless you implement the very same logic, I feel you'll receive garbage in output from the cache.
The usual quick and dirty way to check that this is the case is to exchange the `operator[]` with the `at(const Key&)` one, that will throw if the key you are asking for is not present in the `map`, as I suspect.

Reply:
@rovere ok I agree the filling and getting of the cache should be synchronous, fixed now.
I think this is a good solution for now, since the cache approach is not final.

from @pfs:
"the key based on +z chops the mem in use by 1/2 and it's constructed only once if the caching mode is enabled, the idea was to fold it once more in 60deg sectors in CE-E and 120 deg sectors in CE-H for an additional chopping of the mem. having said that the caching mode is disabled by default as we didn't see much gains in that so far and it should be replaced by a condDB in the future"